pkg/vault/auth.go

```diff
@@ -0,0 +1,202 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package vault
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"strings"
+
+	"github.com/grafana/dskit/flagext"
+	hashivault "github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/api/auth/approle"
+	"github.com/hashicorp/vault/api/auth/kubernetes"
+	"github.com/hashicorp/vault/api/auth/userpass"
+)
+
+const (
+	AppRole    = "approle"
+	Kubernetes = "kubernetes"
+	UserPass   = "userpass"
+	Token      = "token"
+)
+
+type authFactory interface {
+	NewAppRoleAuth(roleID string, secretID *approle.SecretID, opts ...approle.LoginOption) (*approle.AppRoleAuth, error)
+	NewKubernetesAuth(roleName string, opts ...kubernetes.LoginOption) (*kubernetes.KubernetesAuth, error)
+	NewUserpassAuth(username string, password string, opts ...userpass.LoginOption) (*userpass.UserpassAuth, error)
+}
+
+type authMethod interface {
+	authenticate(ctx context.Context, factory authFactory, client *hashivault.Client) (*hashivault.Secret, error)
+}
+
+type AuthConfig struct {
+	AuthType string `yaml:"type"`
+
+	AuthAppRole    AuthAppRole    `yaml:"app_role,omitempty" category:"experimental"`
```

Comment:
we need to put `category:"experimental"` tag onto all config options in these structs, otherwise it's not visible.